### PR TITLE
Implement `#"raw strings"`

### LINKED ISSUE #1121: [Feature request] Raw strings, without backslash escapes or interpolation
String "literals" are not actually literal: the lexer's `readString` routine still expands macro args like `\1` and interpolations like `{x}`, as well as understanding certain backslash escapes like `\n`. (This is done as special handling within `readString` itself; the expansions in `peek` are disabled while reading a string literal, so the expanded contents can be contained as-is inside the string.)

Sometimes it would be more useful not to do this. If you don't need to expand and values within your string, it can be annoying to write `"()[]{}"` and then have to turn it into `"()[]\{}"`. Likewise if you either don't need literal whitespace characters or are using a `"""multi-line string"""` for them, you have to turn `\n` into `\\n`.

This issue proposes "raw strings", like the ones in many programming languages. Syntax could be anything; a few ideas:

- `#"hash strings"` (or `#"""multi-line"""`) would be like Rust.
- `@"at-sign strings"` (or `@"""multi-line"""`) would be like C#.
- `'single-quoted'` strings (or `'''multi-line'''`) would be like PHP and Bash (although we may want to reserve that syntax for possible character literals or something).
- Prefix `r"raw"` like Python could also work, but currently that lexes as `r "raw"`, e.g. passing `"raw"` to macro `r`. `@` and `#` are okay because `@` is always defined as the PC, and `#` is not a valid identifier or operator.
- Any unambiguous punctuation prefix, e.g. `$"dollar strings"`.

(A more niche use case: regular expression literals. `#"\w+[\/]\w+\.\w{3}"` would be like `"\\w+[\\/]\\w+\\.\\w\{3}"`.)

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #1121

This implements the `#"hash prefix"` syntax for raw strings. It can easily be changed to a different prefix; e.g. `@"raw strings"`, or even `$"raw strings"` would be unambiguous too (although a dollar sign conventionally seems like it would *enable* interpolating features, not *disable* them).

### CODE REVIEW

**Comment by aaaaaa123456789** [comment]
If at all possible, I'd rather use a character that is already in use for something else as a prefix, so that unused characters remain available for future features. If you don't like `$`, maybe `%`?

**Comment by Rangi42** [comment]
The main constraint I have is wanting raw string syntax to resemble eventual raw identifier syntax, if we add that.

- Can't do `$"raw"` because `$DEF` is a hex literal not a raw `DEF` identifier
- Can't do `%"raw"` because `%add` is a binary literal (with `opt bad`) not a raw `add` identifier
- Can't do `@"raw"` because `@local` might be an auto-purged local symbol (see discussion in #342)
- But `#"raw"` is okay because `#DEF` works as a raw `DEF` identifier

**Comment by ISSOtm** [comment]
I'm slightly concerned about the use of `#`, regarding potential interop with C tooling. We know that `cpp` has a tendency to generate a bunch of those everywhere (`#line`, anyone?), and so I'm not sure we should use that specific character.

OTOH, I don't see a better alternative, and my concern may be irrelevant in practice. What do you think?

**Comment by Rangi42** [comment]
Hmm. I'm not familiar with the GBDK dev process. Do you have any examples of .asm files that go through it? Would they have preprocessor directives?

We do currently allow `#` in identifiers, so the character can already occur mid-line. And string literals can't be first thing in a line, so this won't change that.

**Comment by ISSOtm** [comment]
GBDK doesn't matter here, as we integrate with SDLD object files. The concern is about other toolchains, such as any LLVM port.

**Comment by Rangi42** [comment]
Okay, then my second paragraph still applies.

(A use case for c preprocessor before assembly: user-`#define`d functions. Although asm comments would be seen by it, so would have to be valid syntax.)

File: man/rgbasm.5 [hunk 1]
```diff
@@ -412,12 +412,21 @@ There are a number of escape sequences you can use within a string:
...
 .Ql \en .
 .Pp
+Raw strings are prefixed by a hash
+.Sq # .
+Inside them, backslashes and braces are treated like regular characters, so they will not be expanded as macro arguments, interpolated symbols, or escape sequences.
+For example, the raw string
+.Ql #"\t\1{s}\"
+is equivalent to the regular string
+.Ql "\\t\\1\{s}\\" .
+(Note that this prevents raw strings from including the double quote character.)
```
Comment (ISSOtm) [inline man/rgbasm.5:427]: The solution I've seen Rust and I *think* Python adopt for this, is that the closing quote must be followed by as many `#`s that preceded the opening one. Should we go with that?

File: test/asm/raw-strings.asm [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+		"new,\nline"
+	test \
+		#"""new,\nline""", \
+		"""new,\\nline"""
+	test \
+		#"/\w+(\+\w+)?@[a-z]+\.[a-z]{2,3}/i", \
+		"/\\w+(\\+\\w+)?@[a-z]+\\.[a-z]\{2,3}/i"
+	test \
+		#{q}{q}{q}rs", \
+		{q}\{q}\{q}rs"
```
Comment (ISSOtm) [inline test/asm/raw-strings.asm:54]: I feel a shiver run down my spine...

File: man/rgbasm.5 (hunk 1 above)
Comment (Rangi42) [reply man/rgbasm.5:429]: I shy away from the idea of putting `#`-counting code in the lexer. As it is, triple-quoted `#"""raw "string"s"""` can contain double quotes, just not three in a row. In the unlikely event that you need triple quotes in a raw string, just use `STRCAT`.

File: test/asm/raw-strings.asm (hunk 2 above)
Comment (Rangi42) [reply test/asm/raw-strings.asm:54]: It's absolutely not code you should write, but if you did, it's a predictable result! :P See also: https://github.com/rust-lang/rust/blob/master/tests/ui/weird-exprs.rs

File: man/rgbasm.5 (hunk 1 above)
Comment (ISSOtm) [reply man/rgbasm.5:429]: Codegen backends may hate us for this, but ok let's go with that.

**Review by ISSOtm — APPROVED** [review]

**Pull request merged**